### ISSUE TITLE
[Java] Checked collections not suggested or recommended as panacea to catch ClassCastException in generics cast to raw types. 

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#1989]
**Affects PMD Version:** 
6.17.0

**Rule:**
No rule as yet. New rule possibly needed. 

**Description:**
Checked collections should be recommended or suggested for code casting or converting to raw types. 

**Code Sample demonstrating the issue:**

```java
package collections;

import static java.util.stream.Collectors.joining;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class FootShootWithSafetyCatch {
  
  private static List<String> names;

  public static void main(String... args) {
    names = Collections.checkedList(new ArrayList<String>(), String.class);
    Collections.addAll(names, "John", "Anton", "Heinz");
    List huh = names;
    List<Integer> numbers = huh;
    numbers.add(42);
    System.out.println(names.stream().collect(joining("+")));
  }
}
```

**Running PMD through:** *
[CLI | Ant ] 

Openjdk 12 on Arch Linux for Termux. 

 The -Xlint:all option in the Java compiler will flag both rawtypes and unchecked errors in the above code sample. 

Do we need a PMD rule that suggests or recommends checkedcollections usage in the scenario above? There are no existing PMD or Checkstyle rules that flag the above (to the best of my knowledge).
This [Reddit post](https://amp.reddit.com/r/java/comments/6ltc94/java_linting/) suggests that the trifecta of PMD, Checkstyle and javac are complementary.

The code sample is borrowed from [here](http://www.javaspecialists.eu/archive/Issue251.html)

**Comment by Fernal73** [comment]
@jsotuyod I don't have a real-world scenario where raw types would actually be needed. Maybe, I was over-enthusiastic. 

There was a case for it when generics were first introduced. 

Perhaps, that's only my lack of experience on projects involving generics mixed with raw types. 

As Java has matured since the introduction of generics, the likelihood is less. 
However, I agree with you that PMD---since it works on class files---must catch all the errors flagged by Xlint:all and provide helpful solutions with examples for every one of those warnings. Additionally, all those SuppressWarnings annotation values supplied by javac must be supported by PMD as well.

**Comment by Fernal73** [comment]
I get the feeling that I'm expected to be aggressive while suggesting such a rule but I can't help remembering that this is an open source project and this isn't something I'll be coding. At least, not yet.

**Comment by Fernal73** [comment]
@jsotuyod that's wonderfully idealistic. But then, FOSS is idealistic.
So does this mean that PMD will now supplement javac's -Xlint:all or do we have to run the linter supplied with javac for better coverage?

**Comment by Fernal73** [comment]
Is there any issue or discussion on PMD where ideas can be thrown up , no matter how ridiculous they appear or how un-implementable they are (within the current API or Framework)?

**Comment by adangel** [comment]
We have various ways: https://pmd.github.io/#support
Most active is for sure this issue tracker on github - meaning here is the highest probability, that someone will respond and join the discussion.
And it's perfectly fine to create an issue for a question or suggestion. We can label these issues accordingly and discuss it via the comments, as we do here.

**Comment by adangel** [comment]
Maybe we could have a rule, that flags the following assignments?

```java
List huh = names; // assigning List<String> to raw List
List<Integer> numbers = huh; // assigning raw List to List<Integer>
```

That's very similar to flagging raw types. We probably should make the rule then universal to not only find List usages, but any generic class assigned to the corresponding raw class.
The rule should then also understand `@SuppressWarnings("rawtypes")` - which simply suppresses it.

The problem in the example is the assignment, which might lead later on to runtime exceptions or undefined/unexpected behavior, since by using raw types, you don't let the compiler check the types at compile time. Using `Collections.checkedList` might help in making you aware of having programmed a bug, but it doesn't prevent it in the first place.

**Comment by Fernal73** [comment]
That's fine with me. The checkedXXX methods must be suggested as ways of
mitigating the effects of these errors in the rule message against types
that match these methods.

As an aside, you have a Rule Designer which allows users to design their
own rules, if needed. Any particular reason why the PMD API documentation
isnt made available as well?

**Comment by Fernal73** [comment]
The API documentation provides another way to check the viability of suggested rules other than the Rule Designer unless you're expecting developers to add to the existing API framework.

**Comment by oowekyala** [comment]
> As an aside, you have a Rule Designer which allows users to design their
own rules, if needed. Any particular reason why the PMD API documentation
isnt made available as well?

Do you mean adding support for javadoc viewing in the designer ? That's on the roadmap.

If you're just talking about our javadocs, then you can find them eg there: https://javadoc.io/doc/net.sourceforge.pmd/pmd-java/6.18.0

**Comment by Fernal73** [comment]
I meant the latter; good to know about the former.

**Comment by Fernal73** [comment]
The documentation has focus set to the Search field on every page. This
scrolls the screen right on every page. Very annoying on a small mobile
device.

**Comment by blerner** [comment]
:+1: to this request -- I would love support for detecting raw types; my students make this mistake often enough that I'd like to turn it into a mandatory error for them...

**Comment by Fernal73** [comment]
That'd be very useful indeed.

It would also come in handy when migrating code or samples from earlier
versions.

**Comment by adangel** [comment]
@blerner : you can compile with the flags `-Xlint:unchecked -Werror` to produce a compiler error, if raw types are used.

**Comment by Fernal73** [comment]
Apparently, you can do the same with priority level of PMD rules.


https://pmd.github.io/latest/pmd_userdocs_configuring_rules.html#message-and-priority-overriding

**Comment by blerner** [comment]
Thanks, yes, I know about `-Xlint:unchecked -Werror`.  The ergonomics of that aren't in my favor though: students don't yet understand how to read those compiler outputs (compared to PMD warnings that I can display to them more clearly), there are other common warnings that I wouldn't want to turn into fatal errors, and I'd like to give students feedback about these particular errors while their code parses-but-doesn't-yet-compile, rather than as a final gatekeeper before they can even run our unit tests.  (Students often make the rational choice to focus on passing our test cases, and then fixing their style errors as a final pass.  By running PMD as a separate component of their grade, I let them parallelize their efforts in improving their code, and anything that acts as a gatekeeper to being able to run unit tests just lengthens their critical path...)

**Comment by Fernal73** [comment]
I suppose it wouldn't be possible to compile these warnings into PMD error
messages, would it?

I run the javac linter myself while building my code.

https://github.com/Fernal73/LearnJava/blob/master/buildj

**Comment by adangel** [comment]
> I suppose it wouldn't be possible to compile these warnings into PMD error
messages, would it?

Sure, but that requires that someone implements a rule - which is this issue about. My suggestion was, to just use the java linter, since there it is already implemented, and not to wait for this rule.

**Comment by Fernal73** [comment]
That's a workaround which I'm fine with. What I'm suggesting is that PMD
adds a feature that runs the javac linter and adds that output to the PMD
violations. This could be switch to be turned on or off.

**Comment by blerner** [comment]
@adangel -- Right.  But since `javac` doesn't produce output in an easily machine-parseable form, I'd have to write another custom tool to programmatically run a `javax.tools.JavaCompiler` and collect the warnings that way...and I was really hoping not to have to write another custom tool :)  It's not that it's inordinately hard, rather just that it takes time to do, and even more time to do robustly enough to use as a drop-in replacement for `javac`...

**Comment by Fernal73** [comment]
I agree. Would Oracle be interested in outputting their warnings in an
easily parseable manner?

**Comment by Fernal73** [comment]
Why parse Oracle's output at all? Take their output and include it in the PMD violation messages. There ought to be a boolean flag to allow/disallow javac linter errors. It's not the end of the world if the messages are incomprehensible to users at first.

**Comment by AbhisekLimbu** [comment]
Hi, I'd like to pick this up based on @adangel  scope from Sep 2019;flagging assignments in both directions (generic-typed value assigned to a raw-typed variable, and vice versa), generalized across all generic types, not just List, and respecting @SuppressWarnings("rawtypes").
Planning to implement this under errorprone, starting from the exact code sample in the issue, then expanding test coverage. Will follow the rule-writing guide and open a draft PR once I have something working.
